test/prototype/test_smoothquant.py

```diff
@@ -13,252 +12,252 @@
-                smoothed_weight, -127, 127, torch.int8
+            # Check output shapes match
+            self.assertEqual(
+                q_out.shape,
+                ref_out.shape,
+                f"Output shapes should match: quantized={q_out.shape}, reference={ref_out.shape}",
             )
-            fq_wei = dequantize_per_channel(qw, w_scales, w_zps, input_dtype)
-
-            # Handle activation quantization based on mode
-            if quant_mode == "static":
-                # activation is quantized per-tensor
-                act_min, act_max = torch.aminmax(smoothed_activation.float())
-                max_val_pos = torch.max(-act_min, act_max)
-                activation_scale = max_val_pos / 127.0
-
-                fq_act = (
-                    torch.quantize_per_tensor(
-                        smoothed_activation.float(),
-                        scale=activation_scale.item(),
-                        zero_point=0,
-                        dtype=torch.qint8,
-                    )
-                    .dequantize()
-                    .to(input_dtype)
-                )
-            else:
-                # activation is quantized per-row (batch * sequence_length)
-                qx, x_scales, x_zps = dynamically_quantize_per_channel(
-                    smoothed_activation.float(), -127, 127, torch.int8
-                )
-                fq_act = dequantize_per_channel(
-                    qx,
-                    x_scales,
-                    x_zps,
-                    input_dtype,
-                )
 
-            # Compute final linear operation
-            reference_out = torch.nn.functional.linear(fq_act, fq_wei, b)
+    def test_observer_insertion(self):
+        """Test that PREPARE step correctly inserts SmoothQuantObservedLinear."""
+
+        m = self.ToyMultiLinearModel(has_bias=False).eval()
+
+        # Before quantization - should be regular Linear
+        self.assertIsInstance(m.linear1, torch.nn.Linear)
+        self.assertNotIsInstance(m.linear1, SmoothQuantObservedLinear)
+
+        # PREPARE step - should insert observers
+        config = SmoothQuantConfig(
+            base_config=Int8DynamicActivationInt8WeightConfig(),
+            step=SmoothQuantStep.PREPARE,
+        )
+        quantize_(m, config)
+
+        # After PREPARE - should be SmoothQuantObservedLinear
+        self.assertIsInstance(m.linear1, SmoothQuantObservedLinear)
+        self.assertTrue(hasattr(m.linear1, "obs"))
+
+        # Test calibration
+        test_data = torch.randn(2, 512)
+        m(test_data)
+
+        # CONVERT step - should produce regular Linear with quantized weights
+        config.step = SmoothQuantStep.CONVERT
+        quantize_(m, config)
+
+        # After CONVERT - should be regular Linear again (but quantized)
+        self.assertIsInstance(m.linear1, torch.nn.Linear)
+        self.assertNotIsInstance(m.linear1, SmoothQuantObservedLinear)
+
+    def test_prepare_for_loading(self):
```

Comment:
to really test this, I think we need to do the following:

1. one path with `PREPARE_FOR_LOADING` as you have done here, let's say `model1`
2. another path with `PREPARE` + `CONVERT`, save the model (torch.save(model2.state_dict())
3. load the state_dict from file, and `model1.load_state_dict(...)`
4. compare the result of `model1(*example_inputs)` (with loaded weights) and `model2(*example_inputs)`, they should match exactly

Reply:
Additionally you can also add one for vllm use pattern, like this: https://github.com/pytorch/ao/blob/568c1932a16ae9f30d48da214a88dc0013e98ed8/test/prototype/test_awq.py#L155, this will make sure that vllm works

Reply:
yeah that's fine but we should at least have some test for functionality with ToyLinear I feel